compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -545,13 +545,36 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         // `[[fn#'b -> impl_trait#'b], [fn#'b -> impl_sized#'b]]`
         //
         // for the opaque type generated on `impl Sized + 'b`, We want the result to be:
-        // impl_sized#'b, so iterating forward is the wrong thing to do.
-        for map in self.generics_def_id_map.iter().rev() {
-            if let Some(r) = map.get(&local_def_id) {
-                debug!("def_id_remapper: remapping from `{local_def_id:?}` to `{r:?}`");
-                local_def_id = *r;
-            } else {
-                debug!("def_id_remapper: no remapping for `{local_def_id:?}` found in map");
+        // impl_sized#'b, so, trying to find first from the start (outermost) would give the wrong result, impl_trait#'b.
+        if let Some(innermost) =
+            self.generics_def_id_map.iter().rev().find_map(|map| map.get(&local_def_id))
+        {
+            local_def_id = *innermost;
+
+            // Once we find the innermost mapping we need to keep remapping forward, so inner
+            // remaps are applied.
+            //
+            // Consider this signature:
+            //
+            // fn test(&self) -> impl Iterator<Item = (&'s usize, impl Iterator<Item = &Number<'s>>)>`
+            //
+            // We would have a '_ anonymous lifetime for &self, that's going to be mapped into the
+            // following inner impl traits.
+            //
+            // When we are processing the innermost impl trait, the one with `Item=&Number<'s>`, we
+            // would end with a generics_def_id_map like:
+            //
+            // `[[self#'_ -> impl_iterator_1#'_], [impl_iterator_1#'_ -> impl_iterator_2#'_]]`
+            //
+            // So starting with self#'_, we want to return impl_iterator_2#'_ and this is why
+            // afterwards we need to keep applying remapping forward.
+            for map in &self.generics_def_id_map {
```

Comment:
Using the same reasoning as the outer loop: why shouldn't this be in reverse order?

Reply:
This is not needed because the mapping would now also be ... `[[self#'_ -> impl_iterator_1#'_], [self#'_ -> impl_iterator_2#'_]]`